rfd/0xxx-partial-cache-healthiness.md

```diff
@@ -0,0 +1,318 @@
+list of rejected kinds on the first `OpInit` event or a failure depending on the supplied AllowRejectedKinds flag. 
+
+
+### Changes to the Events interface
+
+Because the `Events` interface is central to the caching subsystem, it will require some changes. Also, because it has
+so many implementations and is used in so many contexts, some rules need to be established and all the implementations
+must be updated to behave in a consistent way.
+
+None of the proposed changes to the interface are direct though. They all happen through related types and behaviour.
+
+
+#### types.Watch
+```diff
+type Watch struct {
+	Name string
+	Kinds []WatchKind
+	QueueSize int
+	MetricComponent string
++
++	// AllowRejectedKinds enables partial success mode in which the request won't fail if some of the requested Kinds
++	// are not available for watching.
++	AllowRejectedKinds bool
++
++	// KindsRejectedInFlight contains a list of kinds rejected by upstream implementations in a chain of NewWatcher()
++	// calls. Must be empty if AllowRejectedKinds is not set. A kind cannot be on both Kinds and KindsRejectedInFlight
++	// at the same time.
++	KindsRejectedInFlight []WatchKind
+}
+```
+
+The new field `AllowRejectedKinds` should be added to preserve backwards compatibility when a request for an event stream
+crosses service boundary. If an old downstream service unaware of partial success mode connects to an upgraded auth server,
+it won't set this flag. In that case it should get an error if one of the kinds requested for watching is unavailable.
+
+`KindsRejectedInFlight` should be added to accommodate wrapper implementations of `Events` which perform validations
+and call the inner implementation. If `AllowRejectedKinds` is set and the wrapper determines that some of the requested
+kinds cannot be watched, they will be moved from Kinds to this new field and passed to the inner implementation this way.
+
+#### types.WatchStatus
+
+The `Event` struct has fields for a type and an associated resource. Currently, events with type `OpInit` have
+the associated resource field set to `nil`. We should add a new resource type that can be attached to `OpInit` events
+and carry the list of resource kinds what weren't available for watching:
+
+```protobuf
+message WatchStatusV1 {
+  string Kind = 1;
+  string SubKind = 2;
+  string Version = 3;
+  Metadata Metadata = 4;
+  WatchStatusSpecV1 Spec = 5;
+}
+
+message WatchStatusSpecV1 {
+  repeated WatchKind RejectedKinds = 1;
+}
+```
+
+
+#### Rules for `Events` / `NewWatcher()` implementations 
+
+At the beginning of every `NewWatcher(_ context.Context, w types.Watch)` a few invariants must be established:
+1. `w.Kinds` must not be empty.
+2. `w.KindsRejectedInFlight` must be empty unless `w.AllowRejectedKinds` is true.
+3. A kind+subkind cannot be on both `w.Kinds` and `w.KindsRejectedInFlight` at the same time.
```

Comment:
note: in order to prevent concurrency bugs, we'll need to do allocate a new slice for `w.Kinds` each time we remove an element.

Reply:
That was the idea, but it's good to make a note of that https://github.com/gravitational/teleport/pull/23494/commits/9a622b5224908a33dea963271332ba45a03213f7